### ISSUE TITLE
[SUGGESTION] "Play Next" semantics should override repeat once

### ISSUE THREAD

**Issue by JonasT** [vanilla-music/vanilla#891]
- Vanilla Music Version (Settings -> About): 1.0.80
- Android Version: should affect all versions (enhancement request)
- Phone vendor / model: should affect all models (enhancement request)
- Application was downloaded from (use `[x]`):
  - [ ] Play Store
  - [X] F-Droid
  - [ ] Other (Nightly Build)
- Did this work in a previous version?
  - [ ] Yes
  - [X] No
  - [ ] NA / Don't know.

## Description

Currently, when enabling repeat for a single song then playing a single song, the song will repeat forever - good.

But now if while that song is playing, I choose the "Play Next" option on any *other* song, it will be queued up but *never* actually play, the current song keeps repeating. This is the behavior I would totally expect from the "Enqueue" option, but not "Play Next"! In my opinion, the most intuitive behavior for "Play Next" would be to play the chosen song next after whatever plays now finishes *no matter what*, and only after that song change, return to regular behavior (like repeat settings, ...) as usual. Any use of "Play"/"Play Next"/"Play All" or using the track jump buttons could then clear this scheduled forced song change again if it hasn't happened yet, and return to normal behavior as well

## Steps to reproduce the issue

1. Start playing a song
2. Set it to repeat that single song (not playlist repeat, single track repeat!)
3. Choose "Play Next" on any other song, and wait to see that it never actually plays next

**Comment by andiandi13** [comment]
It actually play for me even if the single song repeat is on, it switches to both song every time I tap next or previous.

**Comment by JonasT** [comment]
> See it as a "enqueur next" option.

There is already an "Enqueue" option though, and it's named "Play Next". That is why I am suggesting to make it actually behave as such - unless of course you all disagree with this being more intuitive

**Comment by andiandi13** [comment]
As I just said it's more an enqueue next option. When you do Play all it do a big Play next.

**Comment by JonasT** [comment]
Maybe it should be renamed then?

I would find an actual "Play Next" super useful though - I often listen to songs on repeat, so if I want to have a "smooth" transition to another song on repeat I always need to wait for it to reach the end before I tap stuff which is rather inconvenient.

So my vote would be to actually make it behave as it says, which is actually play it next no matter the current repeat settings. But I understand that I'm not the only one with an interest or use case for this option

**Comment by JonasT** [comment]
More naming / action suggestions:

- Current "Play Next" -> "Enqueue Next"
- Current "Enqueue" -> "Enqueue At End"
- Add a new "Play Next" that actually plays next no matter what

that's kind of what I'd intuitively name things / how I would expect them to work as an end user

**Comment by andiandi13** [comment]
I really don't know how it works, but my expectations are that when _Single repeat_ is on, Vanilla prevent that the rest of the queue plays, no matter that there is one song or more, and that the song came from a "Play Next" action or by a simple enqueue.

So I don't see how Vanilla could do what you ask without turning off single repeat... Just wait for Adrian response I guess

**Comment by JonasT** [comment]
@andiandi13 right, **and I completely agree!** In fact, I propose this behavior to rectify this:

1. Toggling "Single Repeat" in any way, using "Play", "Play All", "Play next", ... **clears out** a previously scheduled forced advance by any previous "Play Next". In that case what you propose is ensured, that is if a user enables "Single Repeat" **after using "Play Next"** it will indeed override and ensure a single repeat as requested

2. However, if "Play Next" is used with single repeat already enabled, it should schedule a forced track advance after the current song finished as I suggested. This ensures that if the user does it in this order, expecting his last "Play Next" request to override previous settings, it will advance to the next song no matter what, which I would strongly suggest is for 99% of users the expected behavior **if the actions were issued in that order**

Do you agree this is the most intuitive behavior for "Play Next"?

Also, **I really don't want to rob you of the current behavior!!!** Which is why I am suggesting to keep it as "Enqueue Next" (and the "Enqueue as "Enqueue At End").

**Comment by adrian-bl** [comment]
I don't think that this should be changed:

As @andiandi13 pointed out: Play next just adds it to the queue after the currently playing song (the action is actually called enqueue_as_next internally).

Having it change the play action (repeat, single, etc) would IMO be very unexpected. So closing this as WONTFIX.

**Comment by JonasT** [comment]
@adrian-bl so why not rename it to "Enqueue As Next" then? If it had been named like that I wouldn't have made the ticket in the first place, because with the current name I also find the current behavior somewhat unexpected
